Andy/three toggle dark mode

WishingWell13 opened the pull request:
<!-- Fill in all spots surrounded by brackets and confirm all check boxes after confirming completion of the tasks -->

# Info

Closes #43.

# Description

What changes did you make? List all distinct problems that this PR addresses. Explain any relevant
motivation or context.

Modified dark/light mode toggle to have a third option for system theme

## Changes

- Made dark/light mdoe toggle into a three way toggle between light mode, system theme, and dark mode

# Type of Change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as
      expected)
- [ ] Logistics Change (A change to a README, description, or dev workflow setup like
      linting/formatting)
- [ ] Continuous Integration Change (Related to deployment steps or continuous integration
      workflows)
- [ ] Other: (Fill In) <!-- Edit this type of change if you select this -->

# Testing

I have tested that my changes fully resolve the linked issue ...

- [x] locally on Desktop.
- [ ] locally on mobile - use https://ngrok.io to get a copy on a mobile device
- [x] on the live deployment preview on Desktop.
- [x] on the live deployment preview on Mobile.
- [x] I have run and passed all new and existing Cypress tests. Add screenshots below.

# Checklist

- [x] I have performed a self-review of my own code.
- [x] I have followed the style guidelines of this project.
- [x] I have documented my code's `src/lib` functions and commented hard to understand areas
      anywhere else.
- [x] My changes produce no new warnings.

# Screenshots

Please include a screenshot of your Cypress testing suite passing successfully.
<img width="378" alt="image" src="https://github.com/acmucsd/membership-portal-ui-v2/assets/54158686/25666bdc-8590-4b26-bf96-e5a151184ae4">


If you made any visual changes to the website, please include relevant screenshots below.
<img width="930" alt="image" src="https://github.com/acmucsd/membership-portal-ui-v2/assets/54158686/6da91b8a-2490-465e-b313-a1218d4ecb0f">

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/54158686/7b6d7ecd-f3d6-458a-ba6a-5accc2213820)

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/54158686/9d7162ca-db17-44fd-9a05-4c6f508dbded)

Vercel: 
<img width="965" alt="image" src="https://github.com/acmucsd/membership-portal-ui-v2/assets/54158686/5f192fb9-b6f8-45b8-bcc2-d929d895ab27">
<img width="167" alt="image" src="https://github.com/acmucsd/membership-portal-ui-v2/assets/54158686/61ee319a-2223-4e0c-91ea-4eade1b5c326">

WishingWell13 committed:
Struggling with moving toggle between three states

WishingWell13 committed:
Indicator moves between three states

Known bugs: Sun icon not loading

WishingWell13 committed:
Component now remembers your selection between reloads

Now avoids "Hydration mismatch" (read here)

https://www.npmjs.com/package/next-themes

WishingWell13 committed:
Icons now are not linked to indicator

Icons stay visible when indicator is not over them, functionality working

WishingWell13 committed:
Removed some magic numbers and comments

WishingWell13 committed:
Shrinked three toggle, linked size to indicator size

Also cleaned up more comments and added SCSS variables

WishingWell13 committed:
Switch now responsive to dark/light mode

The colors of the switch itself

WishingWell13 committed:
Added border to switch and responsiveness to dark/light mode

WishingWell13 committed:
Adjusted top and bottom border of indicator

Known issues:
Toggle is too close to ACM words

WishingWell13 committed:
Reduced gap between elements in Navbar

Reduced gap to make toggle fit in navbar

WishingWell13 committed:
Made indicator lighter when in light mode

farisashai commented on src/components/common/ThreeToggle/index.tsx:
```diff
@@ -0,0 +1,209 @@
+import { useTheme } from 'next-themes';
+import { useEffect, useId, useState } from 'react';
+import MonitorIcon from '../../../../public/assets/icons/monitor.svg';
```
Nit: Use absolute imports for any files not located in the immediate folder.

WishingWell13 replied:
https://medium.com/geekculture/making-life-easier-with-absolute-imports-react-in-javascript-and-typescript-bbdab8a8a3a1

SheepTester replied:
If you look at other files, some of them have imports that start with `import ... from '@/...'` I think that's what he means

farisashai commented on src/components/common/ThreeToggle/index.tsx:
```diff
@@ -0,0 +1,209 @@
+import styles from './style.module.scss';
+
+const ThreeToggle = () => {
+  //   const { theme = 'light', onThemeChange } = props;
```
Nit: Delete any code comments you didn't end up using that aren't documentation.

farisashai commented on src/components/common/ThreeToggle/index.tsx:
```diff
@@ -0,0 +1,209 @@
+    }
+  };
+
+  // const calcIcon = (theme: string | undefined) => {
```
Nit: Delete

farisashai commented on src/components/common/ThreeToggle/index.tsx:
```diff
@@ -0,0 +1,209 @@
+  const iconStyle = calcIconStyle(theme);
+  // const currIcon = calcIcon(theme);
+
+  useEffect(() => {
```
Seeking to Learn: Why is this setState hardcoded into a useEffect that only runs once rather than the initial value for the setState?

WishingWell13 replied:
This was fixing a hydration issue I believe.

https://www.npmjs.com/package/next-themes

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/54158686/f9de6ce5-ad9b-4ed3-98c4-30f79ba63e1a)

SheepTester replied:
I'm not sure how next/themes works but my guess is that the server side doesn't know your color preference (especially if it changed between page sessions) so a difference in color scheme results in the server and client being different, thus hydration error

farisashai commented on src/components/common/ThreeToggle/index.tsx:
```diff
@@ -0,0 +1,209 @@
+      case 'dark':
+        return styles.switchDarkMode;
+      default:
+        console.log(newTheme);
```
Nit: Delete console logs after finishing a feature

WishingWell13 committed:
Fixed depreciation issue with a SCSS variable

Can't do raw division without a calc wrapper

WishingWell13 committed:
Removed unused variables and console statements

WishingWell13 committed:
Resolved linting errors regarding improper units

WishingWell13 committed:
Merge branch 'main' into andy/threeToggleDarkMode

WishingWell13 committed:
Tried to resolve visual errors after merge

Not successful yet

WishingWell13 committed:
Fixed wainbow bar from mobile nav mispositioned

Toggle makes mobile navbar smaller, so needed to lock the height equal to margin top

Wainbow bar now properly is just below normal navbar now (non mobile)

WishingWell13 committed:
Removed comments of old code

WishingWell13 committed:
Merge branch 'main' into andy/threeToggleDarkMode

SheepTester commented on src/components/common/ThreeToggle/index.tsx:
```diff
@@ -0,0 +1,152 @@
+      case 'dark':
+        return 'Icon representing dark theme is on';
+      default:
+        return 'Icon representing dark theme is on';
```
Out of curiosity, why does it default to dark theme? Is this just to satisfy TypeScript?

WishingWell13 replied:
This was mainly to satisfy TypeScript yes. It should never trigger though.

SheepTester commented on src/components/common/ThreeToggle/style.module.scss:
```diff
@@ -0,0 +1,110 @@
+
+  .indicatorDarkMode {
+    background: #9f9f9f;
+  }
```
I see that you have a lot of functions in the component code (eg `calcIndColorStyle`) to figure out what class to apply to change the colors for light/dark mode. An alternate thing you could do, which is used in other parts of the website, is to use CSS variables:

If you look at [themes.scss](https://github.com/acmucsd/membership-portal-ui-v2/blob/main/src/styles/themes.scss), it defines CSS `var()` variables for various colors on the website, using SCSS `$` variables defined in [vars.scss](https://github.com/acmucsd/membership-portal-ui-v2/blob/main/src/styles/vars.scss), and sets them depending on the theme.

For example, for this part, you could do

```scss
// vars.scss
$light-mode-indicator: #b8b8b8;
$dark-mode-indicator: #9f9f9f;

// themes.scss
:root {
  --theme-mode-indicator: #{vars.$light-mode-indicator};
}
[data-theme='dark'] {
  --theme-mode-indicator: #{vars.$dark-mode-indicator};
}

// style.module.scss (this file)
&indicator {
  background: var(--theme-mode-indicator);
}
```

The main benefit of doing it this way is that you don't need to add any JavaScript to set the colors, so you wouldn't need the `calcIndColorStyle` function.

WishingWell13 replied:
Do you have a specific script I could look at for reference?

WishingWell13 replied:
What would I search on google to learn more about this method? Looks like it would need some adjustment to use system theme.

WishingWell13 replied:
Also, if we want to reuse this component, would it be better to keep all its dependencies inside its own file?

WishingWell13 replied:
Updated

SheepTester commented on src/components/layout/DarkModeToggle/index.tsx:
At this point, I think you can just delete `DarkModeToggle` and replace references to it with `ThreeToggle`

SheepTester reviewed (approved):
The animation looks really nice and smooth! I really like it, this is much better than the toggle we currently have 🥰

I like that you defined constants in your CSS. Magic numbers begone 😎

WishingWell13 committed:
Replaced DarkModeToggle with ThreeToggle

WishingWell13 committed:
Shifted handling of toggle coloring to themes.scss

https://www.npmjs.com/package/next-themes?activeTab=readme

Reduces javascript

WishingWell13 committed:
Removed unused functions

Removed unused javascript functions since we switched to using themes.scss

WishingWell13 committed:
Fixed dangling a and removed final unused function

Look at last commit for what to do to add background to switch

WishingWell13 committed:
Removed systemtheme since unused now

farisashai commented on src/components/common/ThreeToggle/index.tsx:
```diff
@@ -0,0 +1,127 @@
+    }
+  };
+
+  // IF YOU WANT TO ADD BACKGROUND TO SWITCH, USE THIS FUNCTION
```
Delete unused comments

farisashai commented on src/components/common/ThreeToggle/index.tsx:
```diff
@@ -0,0 +1,127 @@
+
+  const lightId = `light${useId()}`;
+  const systemId = `system${useId()}`;
+  const darkId = `light${useId()}`;
```
```suggestion
  const darkId = `dark${useId()}`;
```
should this say dark?

farisashai commented on src/components/common/ThreeToggle/index.tsx:
```diff
@@ -0,0 +1,127 @@
+import SunIcon from '../../../../public/assets/icons/sun.svg';
+import styles from './style.module.scss';
+
+const ThreeToggle = () => {
```
i would rename this to something like ThemeToggle for clarity

WishingWell13 replied:
Done

farisashai commented on src/components/common/ThreeToggle/index.tsx:
```diff
@@ -0,0 +1,127 @@
+  //   }
+  // };
+
+  const calcAltText = (theme: string | undefined) => {
```
```suggestion
  const altText = `Icon representing ${theme} theme is on`
```
Based on how you're using the theme variable, it looks like you could rewrite this entire switch as a single variable with a template literal

WishingWell13 replied:
How do I check/test for alt text?

farisashai commented on src/components/common/ThreeToggle/style.module.scss:
```diff
@@ -0,0 +1,97 @@
+    transform: translate3d(calc(#{$switch-start} + #{$switch-gap * 2}), 0, 0);
+  }
+
+  // input[type='radio'] {
```
delete

farisashai reviewed (changes requested):
small nits for code quality

WishingWell13 committed:
Deleted old comments, made import paths absolute

WishingWell13 committed:
Fixed naming error with ids

WishingWell13 committed:
Renamed ThreeToggle to ThemeToggle

WishingWell13 committed:
Changed alt text calculation to template literal

WishingWell13 committed:
Updated absolute path imports

farisashai commented on src/components/common/ThemeToggle/index.tsx:
```diff
@@ -0,0 +1,98 @@
+
+  const iconSize = 30;
+
+  const calcIndicatorStyle = (newTheme: string | undefined) => {
```
```suggestion
  const indicatorText = styles[theme]
```

same comment as the alt text applies here, you don't need an entire switch to select the CSS class name, you can also inline both of them directly in the tag they're used if it's only once

SheepTester replied:
Does this type check? `styles` is not a record of strings because of the style.module.scss.d.ts, and `useTheme`'s returns `theme` of type `string | undefined`, so you'd have to check if it's equal to one of the class names as a literal before being able to do `styles[theme]`.

farisashai replied:
if you type the theme variable explicitly with the options as 

`theme: "light" | "dark" | "system"`
It should pass type checks and be more correct throughout the file.

farisashai reviewed (approved):
looks good to me after this one last comment is addressed

WishingWell13 committed:
Update src/components/common/ThemeToggle/index.tsx

Co-authored-by: Faris Ashai <farisashai@gmail.com>

WishingWell13 committed:
Removed calcIndicatorStyle function

Using maps (Record) and switches instead

farisashai committed:
Merge branch 'main' into andy/threeToggleDarkMode
